cpp/scenario/controllers/include/scenario/controllers/Controller.h

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (C) 2020 Istituto Italiano di Tecnologia (IIT)
+ * All rights reserved.
+ *
+ * This project is dual licensed under LGPL v2.1+ or Apache License.
+ *
+ * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
+ *
+ * This software may be modified and distributed under the terms of the
+ * GNU Lesser General Public License v2.1 or any later version.
+ *
+ * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef SCENARIO_CONTROLLERS_CONTROLLER_H
+#define SCENARIO_CONTROLLERS_CONTROLLER_H
+
+#include "scenario/controllers/References.h"
+
+#include <chrono>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace scenario {
+    namespace controllers {
+        class Controller;
+        class UseScenarioModel;
+        class SetBaseReferences;
+        class SetJointReferences;
+        using ControllerPtr = std::shared_ptr<Controller>;
+        constexpr std::array<double, 3> g = {0, 0, -9.80665};
+    } // namespace controllers
+    namespace gazebo {
+        class Model;
+        using ModelPtr = std::shared_ptr<Model>;
+    } // namespace gazebo
+} // namespace scenario
+
+class scenario::controllers::Controller
+    : public std::enable_shared_from_this<scenario::controllers::Controller>
+{
+public:
+    using StepSize = std::chrono::duration<double>;
+
+    Controller() = default;
+    virtual ~Controller() = default;
+
+    virtual bool initialize() = 0;
+    virtual bool step(const StepSize& dt) = 0;
+    virtual bool terminate() = 0;
+};
+
+class scenario::controllers::UseScenarioModel
+{
+public:
+    UseScenarioModel() = delete;
+    virtual ~UseScenarioModel() = default;
+
+    UseScenarioModel(const gazebo::ModelPtr model)
+        : m_model(model)
+    {}
+
+    virtual bool updateStateFromModel() = 0;
+
+protected:
+    gazebo::ModelPtr m_model;
+};
+
+class scenario::controllers::SetBaseReferences
+{
+public:
+    SetBaseReferences() = default;
+    virtual ~SetBaseReferences() = default;
+
+    virtual bool setBaseReferences(const BaseReferences& jointReferences) = 0;
+};
+
+class scenario::controllers::SetJointReferences
+{
+public:
+    SetJointReferences() = default;
```

Comment:
Maybe we should initialize the controlled joint list in the constructor?

Reply:
Personal opinion: no, interfaces in C++ need to be be pure interfaces to keep complexity manageable.

Reply:
Ok thanks for the feedback. Do you think we should also remove the custom constructor of the `scenario::controllers::UseScenarioModel` few lines above?

Reply:
> 
> 
> Ok thanks for the feedback. Do you think we should also remove the custom constructor of the `scenario::controllers::UseScenarioModel` few lines above?

Not a big problem if we leave it as it is now, but yes I think that decoupling the actual attributes to the interface is more flexible. Let's say that for example  one controller wants to use the `gazebo::ModelPtr` in some other way (for example if for some reason wants to save it as a `std::weak_ptr` to avoid shared_ptr loops) it is not possible with the current  `scenario::controllers::UseScenarioModel`, with limited additional  benefit. By the way, given that you pass shared_ptr around, have you checked that there are no problems of ownership closed loops?

Reply:
> for example if for some reason wants to save it as a std::weak_ptr to avoid shared_ptr loops

This is a good point that is worth supporting, thanks for reporting. Since it's a very small change, I prefer to address it now.

> By the way, given that you pass shared_ptr around, have you checked that there are no problems of ownership closed loops?

In my initial tests I didn't face any of those problems. Controllers get deleted successfully. Actually, there's a problem during destruction, but it's not related to this.

```cpp
// NOTE: we should terminate the controller here, but plugins are not
//       unloaded when the model is removed.
//       All model plugins are deleted when the simulator is destroyed,
//       and there's no more ECM -> we would get segfault.
ControllerRunner::~ControllerRunner() = default;
```

Reply:
Fixed in 67f87b5bba283e18c59591d1052b02673e81d8b1 and a64ba4bbb28519bf7b6a5ecf45cb65994fe6c923